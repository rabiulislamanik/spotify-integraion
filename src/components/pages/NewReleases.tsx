import React, { useContext, useEffect, useState } from 'react';
import {AuthContext} from '../../contexts/AuthContext';
import axios from 'axios'; 
import { IAlbumItem, INewReleaseAlbumResponse } from '../../custom-types/spotifyreponse';
import {newReleaseUrl} from '../../consts/spotifyUrls';
import CardContainer from '../miscellaneous/CardContainer';
import AlbumCard from '../miscellaneous/AlbumCard';


function NewReleases() {
  const authContext = useContext(AuthContext);
  const initialAlbum:IAlbumItem[] = [];
  const [albums, setAlbums]: [IAlbumItem[], (albums: IAlbumItem[]) => void] = useState(initialAlbum);
  const [loading, setLoading]: [boolean, (loading: boolean) => void] = useState<boolean>(true);
  const [error, setError]: [string, (error: string) => void] = useState("");
  const [nextUrl, setNextUrl]: [string|null, (total: string|null) => void] = useState<string|null>(null);
  

  useEffect(() => {
    fetchNewReleases(newReleaseUrl);
  }, []);

  const handleLoadMore = () =>{
    if(!nextUrl)return;
    setLoading(true);
    fetchNewReleases(nextUrl);
  }

  const fetchNewReleases = (url:string)=>{
    axios.get<INewReleaseAlbumResponse>(url,
      {
        headers: {
          'Authorization': `Bearer ${authContext.user?.access_token}`
        }
      }
      )
      .then(response => {
        if(response){
          setAlbums([...albums,...response.data.albums.items]);
          setLoading(false);
          setNextUrl(response.data.albums.next);
        } 
      })
      .catch(error => {
        const errorMessage =
        error.response.status === 401
          ? "Not Authorized"
          : "An unexpected error has occurred";
        setError(errorMessage);
        setLoading(false);
      });
  }
   
  return (
    <div data-testid="new-releases">
      {!loading && albums.length!==0 && (<h1>New Released Albums</h1>)}
      <CardContainer>
      {albums.map((album,i) => (
        <AlbumCard album={album} key={album.name+i}/>
      ))}
      </CardContainer>

      {loading && albums.length===0 &&(<h1>Loading...</h1>)}
      {!loading && error && (<p className="error-message">{error}</p>)}
      {nextUrl && (<button className="load-more" onClick={handleLoadMore}>Load More</button>)}
      {loading && albums.length>0 && (<p className="error-message">{error}</p>)}
    </div>
  );
}

export default NewReleases;