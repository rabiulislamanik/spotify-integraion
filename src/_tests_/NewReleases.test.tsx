import React from 'react';
import renderer from 'react-test-renderer';
import NewReleases from '../components/pages/NewReleases';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, waitFor } from '@testing-library/react';
import axios from 'axios';

jest.mock('axios');

describe('NewRealeases Component',()=>{
    it('renders correctly', () => {
        const tree = renderer.create(<NewReleases />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

