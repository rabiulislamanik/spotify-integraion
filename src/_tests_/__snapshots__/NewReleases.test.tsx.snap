// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewRealeases Component renders correctly 1`] = `
<div
  data-testid="new-releases"
>
  <div
    className="main-container"
  >
    <div
      className="cards-flex-container"
    />
  </div>
  <h1>
    Loading...
  </h1>
</div>
`;
